Chronograf template variables with Flux (#2766)
influxdata/docs-v2: content/chronograf/v1.9/guides/dashboard-template-variables.md

WHERE "host" = ':host:' and time > :dashboardTime
```

#### Flux
In Flux, template variables are stored in a `v` record.
Use dot or bracket notation to reference the variable key inside of the `v` record:

```js
from(bucket: v.bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStart)
  |> filter(fn: (r) => r._field == v["Field key"])
  |> aggregateWindow(every: v.windowPeriod, fn: v.aggregateFunction)
```

### Use template variables in cell titles
To use a template variable to dynamically change the title of a dashboard cell,
use the `:variable-name:` syntax.

For example, if you have a variable named `field` with a selected value of `temp`
and another variable named `location` with a selected value of `San Antonio`,
the following cell title:

```
:temp: data for :location:
```

Would display as:

{{< img-hd src= "/img/chronograf/1-9-template-var-title.png" alt="Use template variables in cell titles" />}}

## Predefined template variables

Chronograf includes predefined template variables controlled by elements in the Chronograf UI.

> These template variables can be used in any of your cells' queries.

Reviewer: ```suggestion
Use predefined template variables in your cell queries.
```

InfluxQL and Flux include their own sets of predefined template variables:

{{< tabs-wrapper >}}
{{% tabs %}}
[InfluxQL](#)
[Flux](#)
{{% /tabs %}}
{{% tab-content %}}
- [`:dashboardTime:`](#dashboardtime)
- [`:upperDashboardTime:`](#upperdashboardtime)

> - [`:autoInterval:`](#autointerval)

Author: ```suggestion
- [`:interval:`](#interval)
```

WHERE time > :dashboardTime: AND time < :upperDashboardTime:
```

> ### autoInterval

Author: ```suggestion
### interval
```

> The `:autoInterval:` template variable is defined by the interval dropdown in the Chronograf dashboard.

Author: ```suggestion
The `:interval:` template variable is defined by the interval dropdown in the Chronograf dashboard.
```

SELECT mean("usage_system") AS "Average System CPU Usage"
FROM "telegraf".."cpu"
WHERE time > :dashboardtime:

> GROUP BY time(:autoInterval:)

Author: ```suggestion
GROUP BY time(:interval:)
```

SELECT mean("usage_system") AS "Average System CPU Usage"
FROM "telegraf".."cpu"
WHERE time > :dashboardtime:
GROUP BY time(:autoInterval:)
```
{{% /tab-content %}}
{{% tab-content %}}
- [`v.timeRangeStart`](#vtimerangestart)
- [`v.timeRangeStop`](#vtimerangestop)
- [`v.windowPeriod`](#vwindowperiod)

### v.timeRangeStart

The `v.timeRangeStart` template variable is controlled by the "time" dropdown in your Chronograf dashboard.

<img src="/img/chronograf/1-6-template-vars-time-dropdown.png" style="width:100%;max-width:549px;" alt="Dashboard time selector"/>

If using relative times, it represents the time offset specified in the dropdown (-5m, -15m, -30m, etc.) and assumes time is relative to "now".
If using absolute times defined by the date picker, `v.timeRangeStart` is populated with lower threshold.

```js
from(bucket: "telegraf/autogen")
  |> range(start: v.timeRangeStart)
  |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
```

{{% note %}}

> To use the date picker to specify a particular time range in the past
> which does not include "now", use `v.timeRangeStart` as the lower limit and
> [`v.timeRangeStop`](#vtimerangestop) as the upper limit.

Reviewer: ```suggestion
To use the date picker to specify a time range in the past
without "now", use `v.timeRangeStart` as the lower limit and
[`v.timeRangeStop`](#vtimerangestop) as the upper limit.
```

{{% /note %}}

### v.timeRangeStop
The `v.timeRangeStop` template variable is defined by the upper time limit specified using the date picker.

<img src="/img/chronograf/1-6-template-vars-date-picker.png" style="width:100%;max-width:762px;" alt="Dashboard date picker"/>

> It will inherit `now()` when using relative time frames or the upper time limit when using absolute time frames.

Reviewer: ```suggestion
For relative time frames, this variable inherits `now()`. For absolute time frames, this variable inherits the upper time limit.
```

> It will inherit `now()` when using relative time frames or the upper time limit when using absolute time frames.

Reviewer: ```suggestion
For relative time frames, this variable inherits `now()`. For absolute time frames, this variable inherits the upper time limit.
```

```js
from(bucket: "telegraf/autogen")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
```

### v.windowPeriod
The `v.windowPeriod` template variable is controlled by the display width of the
dashboard cell and is calculated by the duration of time that each pixel covers.
Use the `v.windowPeriod` variable to downsample data to use only enough points to

> display one point per pixel.

Reviewer: To be sure, is this the sole use case?

Reviewer: Is this accurate:
Use the `v.windowPeriod` variable to limit downsample data to display a maximum of one point per pixel.

> Use the `v.windowPeriod` variable to downsample data to use only enough points to
> display one point per pixel.

Author: ```suggestion
Use the `v.windowPeriod` variable to limit downsample data to display a maximum of one point per pixel.
```

SELECT "purchases" FROM "animals"."autogen"."customers" WHERE "customer" = :customMetaVar:
```

#### InfluxQL meta query variable use cases
Use custom InfluxQL meta query template variables when predefined template variable types aren't able to return the values you want.

### Flux Query
Flux query template variables let you define variable values using Flux queries.
**Variable values are extracted from the `_value` column returned by your Flux query.**

#### Flux query variable use cases

> Flux query template variables are great when the values necessary for your
> variable can't be queried with InfluxQL.

Author: ```suggestion
Flux query template variables are great when the values necessary for your
variable can't be queried with InfluxQL or if you need the flexibility of Flux
to return your desired list of variable values.
```

SELECT "purchases" FROM "animals"."autogen"."customers" WHERE "customer" = :customMetaVar:
```

#### InfluxQL meta query variable use cases

> Custom InfluxQL meta query template variables should be used any time you are pulling values from InfluxDB, but the pre-canned template variable types aren't able to return the desired list of values.

Reviewer: ```suggestion
Use custom InfluxQL meta query template variables when predefined template variable types aren't able to return the values you want.
```

Reviewer: Curious if it makes sense to provide more detail here...or is it obvious when variables won't return values you want?

### Flux Query
Flux query template variables let you define variable values using Flux queries.
**Variable values are extracted from the `_value` column returned by your Flux query.**

#### Flux query variable use cases
Flux query template variables are great when the values necessary for your

> variable can't be queried with InfluxQL.

Reviewer: same

These names are either [predefined variables](#predefined-template-variables) or would
conflict with existing URL query parameters.

- `:database:`
- `:measurement:`
- `:dashboardTime:`
- `:upperDashboardTime:`

> - `:autoInterval:`

Author: ```suggestion
- `:interval:`
```

- `:upper:`
- `:lower:`
- `:zoomedUpper:`
- `:zoomedLower:`
- `:refreshRate:`

## Advanced template variable usage

### Filter template variables with other template variables

> [Custom InfluxQL meta query template variables](#influxQL-meta-query) allow you to filter the array of potential variable values using other existing template variables.

Reviewer: ```suggestion
[Custom InfluxQL meta query template variables](#influxQL-meta-query) let you filter the array of potential variable values using other existing template variables.
```
